### PR TITLE
DDB 884 - Add data entires backend

### PR DESCRIPTION
**Pull request by leighjohnson-friday**
Allows a user to add new measures data if they have the correct permissions
Measures listed and grouped date and filter
Upon submission, copies the lastest entry and creates new entries in the database
Has some simple data validation on the form data and also more in-depth validation on the entity objects to prevent incomplete / mismatch of data types

Temp frontend whilst other story is being worked on

### CODE REVIEW

File: config/development.json [hunk 1]
```diff
@@ -20,6 +20,7 @@
     }
   },
   "features": {
-    "missedMilestones": true
+    "missedMilestones": true,
+    "measureUpload": true
```
Comment (dtaylor7) [inline config/development.json:24]: in my other PR i have just set the default to true and disabled in prod, not sure which is preferred option?

File: pages/data-entry-entity/measure-edit/MeasureEdit.js [hunk 2]
```diff
@@ -155,14 +168,250 @@ class MeasureEdit extends Page {
...
+    return {
+      latest: measuresEntities[measuresEntities.length - 1],
+      grouped: groupedMeasureEntities,
+      fields: measureFields
+    }
+  }
+
+  async getEntitiesToBeCloned(entityIds) {
+    const where = { id: entityIds }
+    const userIsAdmin = this.req.user.roles.map(role => role.name).includes('admin');
```
Comment (dtaylor7) [inline pages/data-entry-entity/measure-edit/MeasureEdit.js:248]: use https://github.com/alphagov/co-eu-transition-dashboard/blob/master/models/user.js#L19 instead

File: pages/data-entry-entity/measure-edit/MeasureEdit.js [hunk 3]
```diff
@@ -155,14 +168,250 @@ class MeasureEdit extends Page {
...
+      parsedEntities
+    }
+  }
+
+  async postRequest(req, res) {
+    if(!this.req.user.isAdmin && !this.res.locals.entitiesUserCanAccess.length) {
+      return res.status(METHOD_NOT_ALLOWED).send('You do not have permisson to access this resource.');
+    }
+
+    if (req.body.type === 'entries') {
```
Comment (dtaylor7) [inline pages/data-entry-entity/measure-edit/MeasureEdit.js:367]: are there multiple forms? if so, in other areas of the code we have used different URLs to decide how to handle the post. i.e. /{url}/entries or /{url}/measure ( to edit metadata ). optional comment

File: pages/data-entry-entity/measure-edit/MeasureEdit.js [hunk 4]
```diff
@@ -155,14 +168,250 @@ class MeasureEdit extends Page {
...
+      return await this.addMeasureEntityData(req.body)
+    }
+
+    return res.redirect(`${this.url}/${this.req.params.metricId}`);
+  }
+
+  async addMeasureEntityData (formData) {
+    const formValidationErrors = await this.validateFormData(formData);
+    if (formValidationErrors.length > 0) {
+      this.req.flash('Missing / invalid form data');
```
Comment (dtaylor7) [inline pages/data-entry-entity/measure-edit/MeasureEdit.js:378]: you could return specific error messages here:
```
this.req.flash(formValidationErrors.map(error => `${error.error} <br>`);
```

File: pages/data-entry-entity/measure-edit/MeasureEdit.js [hunk 5]
```diff
@@ -155,14 +168,250 @@ class MeasureEdit extends Page {
...
+      return this.res.redirect(`${this.url}/${this.req.params.metricId}`);
+    }
+
+    const clonedEntities = await this.getEntitiesToBeCloned(Object.keys(formData.entities))
+    const newEntities = await this.CreateEntitiesFromClonedData(clonedEntities, formData)
```
Comment (dtaylor7) [inline pages/data-entry-entity/measure-edit/MeasureEdit.js:383]: capitalised function name? `CreateEntitiesFromClonedData`

File: pages/data-entry-entity/measure-edit/MeasureEdit.js [hunk 6]
```diff
@@ -155,14 +168,250 @@ class MeasureEdit extends Page {
...
+    const { errors, parsedEntities } = await this.validateEntities(newEntities);
+
+    if (errors.length > 0) {
+      this.req.flash('Error in entity data');
+      return this.res.redirect(`${this.url}/${this.req.params.metricId}`);
```
Comment (dtaylor7) [inline pages/data-entry-entity/measure-edit/MeasureEdit.js:388]: as its used a few times, could put this into a function i.e. `get measureUrl() => ${this.url}/${this.req.params.metricId}`

**Review by dtaylor7 — APPROVED** [review]
Looks good, couple of optional minor comments

File: config/development.json (hunk 1 above)
Comment (leighjohnson-friday) [reply config/development.json]: Happy to use your approach so have removed this for now.

File: pages/data-entry-entity/measure-edit/MeasureEdit.js (hunk 2 above)
Comment (leighjohnson-friday) [reply pages/data-entry-entity/measure-edit/MeasureEdit.js]: Done

File: pages/data-entry-entity/measure-edit/MeasureEdit.js (hunk 3 above)
Comment (leighjohnson-friday) [reply pages/data-entry-entity/measure-edit/MeasureEdit.js:371]: There will be another form for updating other measure data so will look to implement this when doing that ticket.

File: pages/data-entry-entity/measure-edit/MeasureEdit.js (hunk 4 above)
Comment (leighjohnson-friday) [reply pages/data-entry-entity/measure-edit/MeasureEdit.js]: Done

File: pages/data-entry-entity/measure-edit/MeasureEdit.js (hunk 6 above)
Comment (leighjohnson-friday) [reply pages/data-entry-entity/measure-edit/MeasureEdit.js]: Updated

**Review by dtaylor7 — APPROVED** [review]

**Pull request merged**